Upstream-Status: Backport
From 3e83cb2400b15b775e24945aa4a581cf41f6c4cb Mon Sep 17 00:00:00 2001
From: Patrik Flykt <patrik.flykt@linux.intel.com>
Date: Thu, 5 Mar 2015 13:32:55 +0200
Subject: service: Connect ethernet networks when created

Ethernet networks are created once a cable is plugged in, i.e. carrier
is detected. According to the documentation in doc/service-api.txt,
ethernet networks will act as a Connect() D-Bus API call is issued with
the cable plug in.

Fix the code so that all ethernet services are connected at once when
created, provided that the network type is autoconnectable either by
default or by explicitely specifying the technology with the
DefaultAutoConnectTechnologies main.conf variable.

Reported by Mihai Neagu.

diff --git a/src/service.c b/src/service.c
index 4cc645a..a3105fa 100644
--- a/src/service.c
+++ b/src/service.c
@@ -6742,8 +6742,32 @@ struct connman_service * __connman_service_create_from_network(struct connman_ne
 
 	if (service->favorite) {
 		device = connman_network_get_device(service->network);
-		if (device && !connman_device_get_scanning(device))
-			__connman_service_auto_connect(CONNMAN_SERVICE_CONNECT_REASON_AUTO);
+		if (device && !connman_device_get_scanning(device)) {
+
+			switch (service->type) {
+			case CONNMAN_SERVICE_TYPE_UNKNOWN:
+			case CONNMAN_SERVICE_TYPE_SYSTEM:
+			case CONNMAN_SERVICE_TYPE_P2P:
+				break;
+
+			case CONNMAN_SERVICE_TYPE_GADGET:
+			case CONNMAN_SERVICE_TYPE_ETHERNET:
+				if (service->autoconnect) {
+					__connman_service_connect(service,
+						CONNMAN_SERVICE_CONNECT_REASON_AUTO);
+					break;
+				}
+
+				/* fall through */
+			case CONNMAN_SERVICE_TYPE_BLUETOOTH:
+			case CONNMAN_SERVICE_TYPE_GPS:
+			case CONNMAN_SERVICE_TYPE_VPN:
+			case CONNMAN_SERVICE_TYPE_WIFI:
+			case CONNMAN_SERVICE_TYPE_CELLULAR:
+				__connman_service_auto_connect(CONNMAN_SERVICE_CONNECT_REASON_AUTO);
+				break;
+			}
+		}
 	}
 
 	__connman_notifier_service_add(service, service->name);
-- 
cgit v0.10.2


